File: lecture_006(2DArray)/l_009_saddlepoint.cpp
#include<iostream>
#include<vector>
using namespace std;

void display(vector<vector<int>> &arr) {
    for(int i=0; i<arr.size(); i++) {
        for(int j=0; j<arr[0].size(); j++) {
            cout<<arr[i][j]<<" ";
        }
        cout<<endl;
    }
}
void input(vector<vector<int>> &arr) {
    for(int i=0; i<arr.size(); i++) {
        for(int j=0; j<arr[0].size(); j++) {
            cin>>arr[i][j];
        }
    }
    display(arr);
}
void diagonaltraversal(vector<vector<int>> &arr) {
    for(int g=0; g<arr.size(); g++) {
        for(int j=g, i=0; j<arr[0].size(); j++,i++) {
            cout<<arr[i][j]<<endl;
        }
    }
}
void saddlepoint(vector<vector<int>> &arr) {
    for(int i=0; i<arr.size(); i++) {
        int minvalj = 0;
        for(int j=0; j<arr[0].size(); j++) {
            if(arr[i][j]<arr[i][minvalj]) {
                minvalj = j;
            }
        }
        bool flag = true;
        for(int k=0; k<arr.size(); k++) {
            if(arr[k][minvalj]>arr[i][minvalj]) {
                flag = false;
                break;
            }
        }
        if(flag == true) {
            cout<<"Saddle Point: "<<arr[i][minvalj]<<endl;
            return;
        }
    }
    cout<<"Invalid Input"<<endl;
}
int main() {
    int n;
    cin>>n;
    vector<vector<int>>arr(n,vector<int>(n,0));
    input(arr);
    //diagonaltraversal(arr);
    saddlepoint(arr);
}

File: lecture_006(2DArray)/l_009_saddlepoint.java
import java.util.Scanner;
public class l_009_saddlepoint {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        int n = scn.nextInt();
        int[][] arr = new int[n][n];
        input(arr);
        //diagonalTraversal(arr);
        saddlePoint(arr);
    } 
    public static void display(int[][] arr) {
        for(int i=0; i<arr.length; i++) {
            for(int j=0; j<arr[0].length; j++) {
                System.out.print(arr[i][j]+" ");
            }
            System.out.println();
        }
    }
    public static void input(int[][] arr) {
        for(int i=0; i<arr.length; i++) {
            for(int j=0; j<arr[0].length; j++) {
                arr[i][j] = scn.nextInt();
            }
        }
        display(arr);
    }
    public static void diagonalTraversal(int[][] arr) {
        for(int g=0; g<arr.length; g++) {
            for(int j=g, i=0; j<arr[0].length; j++,i++) {
                System.out.println(arr[i][j]);
            }
        }
    }
    public static void saddlePoint(int[][] arr) {
        for(int i=0; i<arr.length; i++) {
            int minvalj = 0;
            for(int j=0; j<arr[0].length; j++) {
                if(arr[i][j]<arr[i][minvalj]) {
                    minvalj = j;
                }
            }
            boolean flag = true;
            for(int k=0; k<arr.length; k++) {
                if(arr[k][minvalj]>arr[i][minvalj]) {
                    flag = false;
                    break;
                } 
            }
            if(flag==true) {
                System.out.println("Saddle Point: "+ arr[i][minvalj]);
                return;
            }
        }
        System.out.println("Invalid input");
    }
}